src/funnel.test.ts

```diff
@@ -0,0 +1,1049 @@
+import { sleep } from "../test/sleep";
+import { constant } from "./constant";
+import { doNothing } from "./doNothing";
+import { funnel } from "./funnel";
+
+// We need some non-trivial duration to use in all our tests, to abstract the
+// actual chosen value we use this UnitOfTime (UT) constant. As long as it is a
+// positive integer, the actual value doesn't matter (but the larger it is,
+// the longer the tests would take to run). The number is in milliseconds.
+const UT = 16;
+
+// We use this reducer to collect arguments from each invocation so we can test
+// what the function was invoked with.
+const ARGS_COLLECTOR = (
+  accumulator: ReadonlyArray<string> | undefined,
+  item: string,
+): ReadonlyArray<string> =>
+  accumulator === undefined ? [item] : [...accumulator, item];
+
+describe("reducer behavior", () => {
+  it("passes the reduced arg to the executor", () => {
+    const mockFn = vi.fn();
+    const foo = funnel(mockFn, {
+      reducer: constant("hello world"),
+      invokedAt: "start",
+      burstCoolDownMs: UT,
+    });
+    foo.call();
+
+    expect(mockFn).toHaveBeenCalledTimes(1);
+    expect(mockFn).toHaveBeenLastCalledWith("hello world");
+  });
+
+  it("reduces call args", async () => {
+    const mockFn = vi.fn((_total: number): void => {
+      /* do nothing */
+    });
```

Comment:
**nit:** i mildly prefer one of these two forms, what do you think?

```ts
    const mockFn = vi.fn() as (total: number) => void;
    const mockFn = vi.fn<(total: number) => void>();
```

Reply:
The form I used works when you also want to provide an actual implementation and are using the mock as a wrapper to track things like called arguments or the number of calls. e.g. 

```ts
let accumulator = 0;
const mockFn = vi.fn((param: number) => accumulator + param);

...

expect(accumulator).toBe(3);
expect(mockFn).toHaveBeenCalledWith([3]);
```

That can't be solved with either of the other forms, so it would create two different syntaxes for using vi.fn, depending on whether you need an actual implementation. Do you think that's still better?

Reply:
i think it's a matter of preference. i mildly prefer using one form for empty functions and one form for functions with an implementation, but honestly this is a nit